app/web_conn.go

```diff
@@ -122,43 +128,45 @@ func (wc *WebConn) SetSession(v *model.Session) {
 // Pump starts the WebConn instance. After this, the websocket
 // is ready to send/receive messages.
 func (wc *WebConn) Pump() {
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		wc.writePump()
-	}()
-	wc.readPump()
-	close(wc.endWritePump)
-	wg.Wait()
+	wc.writePump()
 	wc.App.HubUnregister(wc)
 	close(wc.pumpFinished)
 }
 
-func (wc *WebConn) readPump() {
-	defer func() {
-		wc.WebSocket.Close()
-	}()
-	wc.WebSocket.SetReadLimit(model.SOCKET_MAX_MESSAGE_SIZE_KB)
+// ReadMsg will read a single message from the websocket connection.
+func (wc *WebConn) ReadMsg() error {
+	// TODO(agniva):
+	// wc.WebSocket.SetReadLimit(model.SOCKET_MAX_MESSAGE_SIZE_KB)
 	wc.WebSocket.SetReadDeadline(time.Now().Add(pongWaitTime))
-	wc.WebSocket.SetPongHandler(func(string) error {
-		wc.WebSocket.SetReadDeadline(time.Now().Add(pongWaitTime))
+
+	r := wsutil.NewReader(wc.WebSocket, ws.StateServerSide)
+	decoder := json.NewDecoder(r)
+
+	var req model.WebSocketRequest
+	hdr, err := r.NextFrame()
+	if err != nil {
+		return errors.Wrap(err, "error while getting the next websocket frame")
+	}
+	switch hdr.OpCode {
+	case ws.OpClose:
+		// Return if closed.
+		return nil
+	case ws.OpPong:
+		// Handle pongs
 		if wc.IsAuthenticated() {
 			wc.App.Srv().Go(func() {
 				wc.App.SetStatusAwayIfNeeded(wc.UserId, false)
 			})
 		}
-		return nil
-	})
-
-	for {
-		var req model.WebSocketRequest
-		if err := wc.WebSocket.ReadJSON(&req); err != nil {
-			wc.logSocketErr("websocket.read", err)
-			return
+	default:
+		// Default case of data message.
+		if err := decoder.Decode(&req); err != nil {
+			return errors.Wrap(err, "error during decoding websocket message")
```

Comment:
Are there cases where we would need to call [`r.Discard()`](https://godoc.org/github.com/gobwas/ws/wsutil#Reader.Discard)? In other words, can the decoder return error having failed to read the whole message content?

Reply:
Good point. :+1:

Reply:
Done